presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -619,6 +649,44 @@ protected void internalRefreshStats()
         }
     }
 
+    protected void enforceHardMemoryLimit()
+    {
+        checkState(Thread.holdsLock(root), "Must hold lock to enforce hard memory limit");
+        synchronized (root) {
+            if (subGroups.isEmpty()) {
+                if (cachedMemoryUsageBytes > hardMemoryLimitBytes) {
+                    AtomicLong memoryReclaimed = new AtomicLong();
+                    // Kill queries which exceed the hard memory limit
+                    ImmutableList.copyOf(runningQueries).stream()
+                            .filter(queryExecution -> queryExecution.getTotalMemoryReservation() > hardMemoryLimitBytes)
+                            .forEach(queryExecution -> {
+                                memoryReclaimed.addAndGet(queryExecution.getTotalMemoryReservation());
+                                queryExecution.cancelQuery();
```

Comment:
This is a big change, and will have to be discussed. There is no pre-existing code that cancels running queries. In any case, the query should be killed with a specific exception message, not canceled willy-nilly.